src/trio/_core/_run.py

```diff
@@ -883,16 +918,28 @@ def shield(self, new_value: bool) -> None:
             self._cancel_status.recalculate()
 
     @enable_ki_protection
-    def cancel(self) -> None:
+    def cancel(self, reason: str | None = None) -> None:
         """Cancels this scope immediately.
 
+        The optional ``reason`` argument accepts a string, which will be attached to
+        any resulting :exc:`Cancelled` exception to help you understand where that
+        cancellation is coming from and why it happened.
+
         This method is idempotent, i.e., if the scope was already
         cancelled then this method silently does nothing.
         """
         if self._cancel_called:
             return
         with self._might_change_registered_deadline():
             self._cancel_called = True
+        if self._cancel_reason is None:
```

Comment:
Should we set `source_task` outside of this `if` statement, because I see some other `_cancel_reason = CancelReason(...)` calls don't have it.

Reply:
no this one is critically important. Several other cancellations internally use `cs.cancel()`, so if this one is not guarded by `if` then a lot of other cancel reasons will always get overridden.

What might be more sketchy is others not having it, I should maybe add a setter or helper function to `CancelScope`

Reply:
actually, what makes more sense is having an internal `def _cancel` because we should only ever set `_cancel_reason` when doing a cancellation

Reply:
> Several other cancellations internally use `cs.cancel()`, so if this one is not guarded by `if` then a lot of other cancel reasons will always get overridden.

I'm not entirely convinced avoiding this is a good thing:

```py
with trio.move_on_after(1):
  with trio.CancelScope() as cs:
    try:
      try:
        await trio.sleep(5)
      except trio.Cancelled as c1:
        cs.cancel("shutting down")
        await trio.sleep(1)
    except trio.Cancelled as c2:
      # what should c2's reason be?
```

This might matter for instance in code for shutting down stuff on exceptions, moving on after 2 seconds. The cancel reason if the clean up code ran over its 2 seconds would presumably (?) be that the exceptions happened, not that the timeout happened.